examples/pi-home-router/src/arp/arp_generator.rs

```diff
@@ -0,0 +1,114 @@
+use crate::arp::ArpOp::Request;
+use crate::arp::{ArpFrame, ArpTable, IPV4_PROTOCOL_TYPE};
+use crate::types::InterfaceAnnotated;
+use route_rs_packets::{EthernetFrame, Ipv4Packet, Ipv6Packet, MacAddr};
+use route_rs_runtime::processor::Processor;
+use std::convert::TryFrom;
+use std::net::IpAddr;
+
+pub(crate) struct ArpGenerator {
+    // Mapping between (protocol type, protocol address) -> 48-bit Ethernet address
+    arp_table: ArpTable,
+}
+
+impl ArpGenerator {
+    pub fn new() -> Self {
+        ArpGenerator {
+            arp_table: ArpTable::new(),
+        }
+    }
+}
+
+impl Processor for ArpGenerator {
+    type Input = InterfaceAnnotated<EthernetFrame>;
+    type Output = InterfaceAnnotated<EthernetFrame>;
+
+    ///
+    /// As a packet is sent down through the network layers, routing determines the protocol address
+    /// of the next hop for the packet and on which piece of hardware it expects to find the station
+    /// with the immediate target protocol address. The Address Resolution module converts the
+    /// <protocol type, target protocol address> pair to a 48.bit Ethernet address. The Address
+    /// Resolution module tries to find this pair in a table. If it finds the pair, it gives the
+    /// corresponding 48.bit Ethernet address back to the caller which then transmits the packet.
+    ///
+    /// If the lookup for the (protocol type, protocol address) fails, it probably informs the
+    /// caller that it is throwing the packet away (on the assumption the packet will be
+    /// retransmitted by a higher network layer), and generates an Ethernet packet with a type field
+    /// of ether_type$ADDRESS_RESOLUTION. The Address Resolution module then sets
+    ///
+    /// Hardware Type: ares_hrd$Ethernet = 1
+    /// Protocol Type: type that is being resolved
+    /// Hardware Address Length: 6 (the number of bytes in a 48.bit Ethernet address)
+    /// Protocol Address Length: length of an address in that protocol
+    /// Op: ares_op$REQUEST = 1
+    /// Sender Hardware Address: the 48.bit ethernet address of itself
+    /// Sender Protocol Address: the protocol address of itself
+    /// Target Protocol Address: the protocol address of the machine that is trying to be accessed.
+    ///
+    /// It does not set Target Hardware Address to anything in particular, because it is this value
+    /// that it is trying to determine. It could set the Target Hardware Address to the broadcast
+    /// address (FF:FF:FF:FF:FF:FF) for the hardware (all ones in the case of the 10Mbit Ethernet)
+    /// if that makes it convenient for some aspect of the implementation.
+    ///
+    /// It then causes this packet to be broadcast to all stations on the Ethernet cable originally
+    /// determined by the routing mechanism.
+    ///
+    fn process(&mut self, packet: Self::Input) -> Option<Self::Output> {
+        // It would be nice if there was a function to determine which network protocol type
```

Comment:
You should be able to match on `frame.ether_type()`. Probably we could add some human-readable names in router-rs-packets like we do for `IpProtocol`.

Reply:
That's a good point. I defined constants for those in `arp_utils`, but I think they should live in the packet library.